server/src/main/java/org/elasticsearch/index/engine/ThreadPoolMergeExecutorService.java

```diff
@@ -127,13 +137,21 @@ boolean submitMergeTask(MergeTask mergeTask) {
                 );
             }
             // then enqueue the merge task proper
-            queuedMergeTasks.add(mergeTask);
+            enqueueMergeTask(mergeTask);
             return true;
         }
     }
 
     void reEnqueueBackloggedMergeTask(MergeTask mergeTask) {
-        queuedMergeTasks.add(mergeTask);
+        enqueueMergeTask(mergeTask);
+    }
+
+    private void enqueueMergeTask(MergeTask mergeTask) {
+        synchronized (mergeEventsMutex) {
+            if (queuedMergeTasks.add(mergeTask)) {
```

Comment:
I think we can assert that it always returns true instead, thus avoid the if-statement, invoke the listener before adding and thus avoid the synchronized?

Reply:
Sure. See https://github.com/elastic/elasticsearch/pull/126667/commits/ba7ebb55d1931834b7f21a7219c69f6678835298. This doesn't change anything in the stateless PR though.